### PR TITLE
Disallow float offset type in cub::segmented_reducde

### LINKED ISSUE #1221: [BUG]: Improve error reporting for segmented reduce
### Is this a duplicate?

- [X] I confirmed there appear to be no [duplicate issues](https://github.com/NVIDIA/cccl/issues) for this bug and that I agree to the [Code of Conduct](CODE_OF_CONDUCT.md)

### Type of Bug

Compile-time Error

### Component

CUB

### Describe the bug

As of https://github.com/NVIDIA/cccl/pull/414 we use common value type of offset begin / end iterators as offset type in segmented reduce. Incorrect use of API is hard to diagnose. For instance, user can pass offset iterator with `float` as its value type. Float is not expected on the implementation side of segmented reduce, leading to many errors that are hard to associate with API misuse. 

### How to Reproduce

```cpp
#include <cub/cub.cuh>

int main() {
  using offset_t = float;  // error
  // using offset_t = int; // ok
  float *d_in{}, *d_out{};
  offset_t *d_begin_offsets{}, *d_end_offsets{};
  std::size_t temp_storage_bytes{};
  std::uint8_t *d_temp_storage{};
  cub::DeviceSegmentedReduce::Sum(d_temp_storage, temp_storage_bytes, d_in, d_out, 0, d_begin_offsets, d_end_offsets);
}
```

### Expected behavior

We should use combination of SFINAE and static asserts like we do in device radix sort to improve error reporting. 

### Reproduction link

_No response_

### Operating System

_No response_

### nvidia-smi output

_No response_

### NVCC version

_No response_

### PR DESCRIPTION
**Pull request by gonidelis**
## Description

Fixes #1221 for introducing tag_dispatching intermediate layer in cub::SegmentedReduce that improves error reporting and enforces that the offset type be integral.

## Checklist

- [x] Add tests.
- [x] Export Doxygen examples in standalone test cases.

### CODE REVIEW

File: cub/cub/device/device_segmented_reduce.cuh [hunk 1]
```diff
@@ -378,41 +444,38 @@ struct DeviceSegmentedReduce
...
-                                                       num_segments,
-                                                       d_begin_offsets,
-                                                       d_end_offsets,
-                                                       cub::Sum(),
-                                                       OutputT(), // zero-initialize
-                                                       stream);
+    static_assert(integral_offset_check::value, "Offset iterator type should be integral.");
+
+    return custom_segmented_reduce<InputIteratorT, OutputIteratorT, BeginOffsetIteratorT, EndOffsetIteratorT, OffsetT>(
+      integral_offset_check{}, // feels like integral_offset_check::type{}
```
Comment (gonidelis) [inline cub/cub/device/device_segmented_reduce.cuh:468]: Will remove this comment in the next push.

File: cub/cub/device/device_segmented_reduce.cuh [hunk 2]
```diff
@@ -34,6 +34,10 @@
 
 #include <cub/config.cuh>
 
+#include "cuda/std/detail/libcxx/include/__iterator/readable_traits.h"
+#include "cuda/std/detail/libcxx/include/__type_traits/integral_constant.h"
+#include "cuda/std/detail/libcxx/include/__type_traits/is_integral.h"
```
Comment (miscco) [inline cub/cub/device/device_segmented_reduce.cuh:39]: I believe for now we went with the idea that cub should mostly include the public headers from libcu++

We might change that but until then I would recommend to include just `<cuda/std/type_traits>` and `<cuda/std/iterator>`

Also those headers are probably added by clangd because they are in front of our system header guard. We want those to be after it

File: cub/cub/device/device_segmented_reduce.cuh [hunk 3]
```diff
@@ -208,35 +273,36 @@ struct DeviceSegmentedReduce
...
-                                                           num_segments,
-                                                           d_begin_offsets,
-                                                           d_end_offsets,
-                                                           reduction_op,
-                                                           initial_value,
-                                                           stream);
+    static_assert(integral_offset_check::value, "Offset iterator type should be integral.");
+
+    return custom_segmented_reduce<InputIteratorT, OutputIteratorT, BeginOffsetIteratorT, EndOffsetIteratorT, OffsetT>(
+      integral_offset_check{}, // feels like integral_offset_check::type{}
```
Comment (miscco) [inline cub/cub/device/device_segmented_reduce.cuh:295]: I am confused by this. 

We have a static assert in front so there should never be a case where this is passed with `false_type`

File: cub/cub/device/device_segmented_reduce.cuh [hunk 4]
```diff
@@ -73,6 +78,66 @@ CUB_NAMESPACE_BEGIN
...
+
+  template <typename InputIteratorT,
+            typename OutputIteratorT,
+            typename BeginOffsetIteratorT,
+            typename EndOffsetIteratorT,
+            typename OffsetT,
+            typename ReductionOpT,
+            typename InitT>
+  CUB_RUNTIME_FUNCTION static cudaError_t custom_segmented_reduce(
+    ::cuda::std::true_type,
```
Comment (miscco) [inline cub/cub/device/device_segmented_reduce.cuh:110]: I believe that one can go. By adding the internal static assert we are already prohibiting any instantiations of `custom_segmented_reduce` where this is `false_type`

Please consider removing that extra function and just using the static_assert

**Review by miscco — CHANGES_REQUESTED** [review]

File: cub/cub/device/device_segmented_reduce.cuh (hunk 3 above)
Comment (gevtushenko) [reply cub/cub/device/device_segmented_reduce.cuh:295]: @miscco I'm not sure what you mean. Static asserts don't prohibit instantiations of subsequent templates: https://godbolt.org/z/Td1Y35f9r.  There's a case when false type is passed. It happens when user accidentally misuse API. The tag dispatch approach hides pages of underlying errors and instead leads to a single error coming from the front static assert.

File: cub/cub/device/device_segmented_reduce.cuh [hunk 5]
```diff
@@ -73,6 +78,66 @@ CUB_NAMESPACE_BEGIN
...
 {
+private:
+  template <typename InputIteratorT,
+            typename OutputIteratorT,
+            typename BeginOffsetIteratorT,
+            typename EndOffsetIteratorT,
+            typename OffsetT,
+            typename ReductionOpT,
+            typename InitT>
+  CUB_RUNTIME_FUNCTION static cudaError_t custom_segmented_reduce(
```
Comment (gevtushenko) [inline cub/cub/device/device_segmented_reduce.cuh:89]: suggestion: I believe `custom_` was used in the radix sort code to denote custom types support (user-defined). This is not related to the check we have in segmented sort. I'd go with something like:
```suggestion
  CUB_RUNTIME_FUNCTION static cudaError_t segmented_reduce(
```

File: cub/cub/device/device_segmented_reduce.cuh [hunk 6]
```diff
@@ -108,7 +173,7 @@ struct DeviceSegmentedReduce
   //!    struct CustomMin
   //!    {
   //!        template <typename T>
-  //!        CUB_RUNTIME_FUNCTION __forceinline__
+  //!        __device__ __forceinline__
```
Comment (gevtushenko) [inline cub/cub/device/device_segmented_reduce.cuh:176]: praise: good catch!

File: cub/cub/device/device_segmented_reduce.cuh [hunk 7]
```diff
@@ -561,26 +624,22 @@ struct DeviceSegmentedReduce
...
+      d_temp_storage,
+      temp_storage_bytes,
+      d_in,
+      d_out,
+      num_segments,
+      d_begin_offsets,
+      d_end_offsets,
+      cub::Min(),
+      Traits<InputT>::Max(),
+      stream);
```
Comment (gevtushenko) [inline cub/cub/device/device_segmented_reduce.cuh:642]: important: the TODO associated with `Traits<InputT>::Max()` is still valid. We have to replace it with numeric limits at some point in the future. Please, return the comment here and in the related places.

File: cub/cub/device/device_segmented_reduce.cuh [hunk 8]
```diff
@@ -208,35 +273,36 @@ struct DeviceSegmentedReduce
...
-                                                           temp_storage_bytes,
-                                                           d_in,
-                                                           d_out,
-                                                           num_segments,
-                                                           d_begin_offsets,
-                                                           d_end_offsets,
-                                                           reduction_op,
-                                                           initial_value,
-                                                           stream);
+    static_assert(integral_offset_check::value, "Offset iterator type should be integral.");
```
Comment (gevtushenko) [inline cub/cub/device/device_segmented_reduce.cuh:292]: suggestion: I'm not sure about this wording. The offset iterator type can be anything, starting from a `int*`, ending with `thrust::counting_iterator`. Is think it should be something along the lines of `Offset type should be integral`, or `Offset iterator value type should be integral`.

File: cub/test/test_device_segmented_reduce_offset_type_fail.cu [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+  cub::DeviceSegmentedReduce::Min(d_temp_storage, temp_storage_bytes, d_in, d_out, 0, d_offsets, d_offsets + 1);
+
+  // expected-error {{"Offset iterator type should be integral."}}
+  cub::DeviceSegmentedReduce::ArgMin(d_temp_storage, temp_storage_bytes, d_in, d_out, 0, d_offsets, d_offsets + 1);
+
+  // expected-error {{"Offset iterator type should be integral."}}
+  cub::DeviceSegmentedReduce::Max(d_temp_storage, temp_storage_bytes, d_in, d_out, 0, d_offsets, d_offsets + 1);
+
+  // expected-error {{"Offset iterator type should be integral."}}
+  cub::DeviceSegmentedReduce::ArgMax(d_temp_storage, temp_storage_bytes, d_in, d_out, 0, d_offsets, d_offsets + 1);
```
Comment (gevtushenko) [inline cub/test/test_device_segmented_reduce_offset_type_fail.cu:29]: important: currently, this works as follows: CMake builds this targets, searches for the regex in the `{{ ... }}`, and treats the test as passed if the compilation output contains the regex. As written, all API would appear to produce an error even if only a single one of those actually produced it. I don't know a better solution than splitting this into five standalone tests.

File: cub/cub/device/device_segmented_reduce.cuh (hunk 3 above)
Comment (miscco) [reply cub/cub/device/device_segmented_reduce.cuh:295]: Did I totally miss the `false_type` declaration directly above.

I do not believe this is a question for this PR and I definitely wont block it, but I am not sure whether tag dispatch paired with an unimplemented method is a better fit than traditional SNIFAE / concepts.

File: cub/cub/device/device_segmented_reduce.cuh [hunk 10]
```diff
@@ -378,41 +440,38 @@ struct DeviceSegmentedReduce
...
+  template <typename InputIteratorT, typename OutputIteratorT, typename BeginOffsetIteratorT, typename EndOffsetIteratorT>
+  CUB_RUNTIME_FUNCTION static cudaError_t
+  Sum(void* d_temp_storage,
+      size_t& temp_storage_bytes,
+      InputIteratorT d_in,
+      OutputIteratorT d_out,
+      int num_segments,
+      BeginOffsetIteratorT d_begin_offsets,
+      EndOffsetIteratorT d_end_offsets,
+      cudaStream_t stream = nullptr)
```
Comment (gevtushenko) [inline cub/cub/device/device_segmented_reduce.cuh:452]:
```suggestion
      cudaStream_t stream = 0)
```

File: cub/cub/device/device_segmented_reduce.cuh [hunk 11]
```diff
@@ -543,44 +602,41 @@ struct DeviceSegmentedReduce
...
+  template <typename InputIteratorT, typename OutputIteratorT, typename BeginOffsetIteratorT, typename EndOffsetIteratorT>
+  CUB_RUNTIME_FUNCTION static cudaError_t
+  Min(void* d_temp_storage,
+      size_t& temp_storage_bytes,
+      InputIteratorT d_in,
+      OutputIteratorT d_out,
+      int num_segments,
+      BeginOffsetIteratorT d_begin_offsets,
+      EndOffsetIteratorT d_end_offsets,
+      cudaStream_t stream = nullptr)
```
Comment (gevtushenko) [inline cub/cub/device/device_segmented_reduce.cuh:614]:
```suggestion
      cudaStream_t stream = 0)
```

File: cub/cub/device/device_segmented_reduce.cuh [hunk 12]
```diff
@@ -721,18 +777,16 @@ struct DeviceSegmentedReduce
...
+  template <typename InputIteratorT, typename OutputIteratorT, typename BeginOffsetIteratorT, typename EndOffsetIteratorT>
+  CUB_RUNTIME_FUNCTION static cudaError_t ArgMin(
+    void* d_temp_storage,
+    size_t& temp_storage_bytes,
+    InputIteratorT d_in,
+    OutputIteratorT d_out,
+    int num_segments,
+    BeginOffsetIteratorT d_begin_offsets,
+    EndOffsetIteratorT d_end_offsets,
+    cudaStream_t stream = nullptr)
```
Comment (gevtushenko) [inline cub/cub/device/device_segmented_reduce.cuh:789]:
```suggestion
    cudaStream_t stream = 0)
```

File: cub/cub/device/device_segmented_reduce.cuh [hunk 13]
```diff
@@ -909,44 +961,41 @@ struct DeviceSegmentedReduce
...
+  template <typename InputIteratorT, typename OutputIteratorT, typename BeginOffsetIteratorT, typename EndOffsetIteratorT>
+  CUB_RUNTIME_FUNCTION static cudaError_t
+  Max(void* d_temp_storage,
+      size_t& temp_storage_bytes,
+      InputIteratorT d_in,
+      OutputIteratorT d_out,
+      int num_segments,
+      BeginOffsetIteratorT d_begin_offsets,
+      EndOffsetIteratorT d_end_offsets,
+      cudaStream_t stream = nullptr)
```
Comment (gevtushenko) [inline cub/cub/device/device_segmented_reduce.cuh:973]:
```suggestion
      cudaStream_t stream = 0)
```

File: cub/cub/device/device_segmented_reduce.cuh [hunk 14]
```diff
@@ -1090,18 +1139,16 @@ struct DeviceSegmentedReduce
...
+  template <typename InputIteratorT, typename OutputIteratorT, typename BeginOffsetIteratorT, typename EndOffsetIteratorT>
+  CUB_RUNTIME_FUNCTION static cudaError_t ArgMax(
+    void* d_temp_storage,
+    size_t& temp_storage_bytes,
+    InputIteratorT d_in,
+    OutputIteratorT d_out,
+    int num_segments,
+    BeginOffsetIteratorT d_begin_offsets,
+    EndOffsetIteratorT d_end_offsets,
+    cudaStream_t stream = nullptr)
```
Comment (gevtushenko) [inline cub/cub/device/device_segmented_reduce.cuh:1151]:
```suggestion
    cudaStream_t stream = 0)
```

File: cub/cub/device/device_segmented_reduce.cuh [hunk 15]
```diff
@@ -1131,23 +1178,21 @@ struct DeviceSegmentedReduce
...
-                                                     num_segments,
-                                                     d_begin_offsets,
-                                                     d_end_offsets,
-                                                     cub::ArgMax(),
-                                                     initial_value,
-                                                     stream);
+    using integral_offset_check = ::cuda::std::is_integral<OffsetT>;
+    static_assert(integral_offset_check::value, "Offset iterator value type should be integral.");
+
+    return segmented_reduce<ArgIndexInputIteratorT, OutputIteratorT, BeginOffsetIteratorT, EndOffsetIteratorT, OffsetT>(
```
Comment (gevtushenko) [inline cub/cub/device/device_segmented_reduce.cuh:1184]: important: we used to explicitly specify `AccumT`. Now, there's a logic that relies on return type of reduction operator. To be on the safe side, let's pass all of the template arguments as we used to do: reduction op, init value type, accumulator type. Same in other places.

File: cub/cub/device/device_segmented_reduce.cuh [hunk 16]
```diff
@@ -99,47 +164,11 @@ struct DeviceSegmentedReduce
...
-  //!
-  //!    // CustomMin functor
-  //!    struct CustomMin
-  //!    {
-  //!        template <typename T>
-  //!        CUB_RUNTIME_FUNCTION __forceinline__
-  //!        T operator()(const T &a, const T &b) const {
-  //!            return (b < a) ? b : a;
-  //!        }
-  //!    };
```
Comment (gevtushenko) [inline cub/cub/device/device_segmented_reduce.cuh:115]: question: we used to have `CustomMin` example in the doc, but now its definition is out of the literal include scope. Do you think it's useful to have another literal include to provide user with an example of function object?

File: cub/cub/device/device_segmented_reduce.cuh [hunk 17]
```diff
@@ -464,35 +467,17 @@ struct DeviceSegmentedReduce
...
-  //!    void     *d_temp_storage = NULL;
-  //!    size_t   temp_storage_bytes = 0;
-  //!    cub::DeviceSegmentedReduce::Min(
-  //!      d_temp_storage, temp_storage_bytes, d_in, d_out,
-  //!      num_segments, d_offsets, d_offsets + 1);
+  //! .. literalinclude:: ../../test/catch2_test_device_segmented_reduce_api.cu
+  //!     :language: c++
+  //!     :dedent:
+  //!     :start-after: example-begin segmented-reduce-custommin
+  //!     :end-before: example-end segmented-reduce-custommin
```
Comment (gevtushenko) [inline cub/cub/device/device_segmented_reduce.cuh:474]: important: this should've been in the `cub::DeviceSegmentedReduce::Reduce` docs instead of `cub::DeviceSegmentedReduce::Min`.

Comment (gonidelis) [reply cub/cub/device/device_segmented_reduce.cuh:474]: excellent catch

**Review by gonidelis — COMMENTED** [review]
Move custommin code literal beyond cub::DeviceSegmentedReduce::Reduce

**Review by gevtushenko — APPROVED** [review]

**Comment by gonidelis** [comment]
@miscco merge?

**Review by miscco — APPROVED** [review]

File: cub/test/test_device_radix_sort_decomposer_fail.cu [hunk 18]
```diff
@@ -1,3 +1,30 @@
+/******************************************************************************
+ * Copyright (c) 2011-2022, NVIDIA CORPORATION.  All rights reserved.
```
Comment (miscco) [inline cub/test/test_device_radix_sort_decomposer_fail.cu:2]:
```suggestion
 * Copyright (c) 2011-2024, NVIDIA CORPORATION.  All rights reserved.
```

File: cub/test/test_device_segmented_reduce_offset_type_fail.cu [hunk 19]
```diff
@@ -0,0 +1,66 @@
+/******************************************************************************
+ * Copyright (c) 2011-2022, NVIDIA CORPORATION.  All rights reserved.
```
Comment (miscco) [inline cub/test/test_device_segmented_reduce_offset_type_fail.cu:2]:
```suggestion
 * Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
```

**Review by miscco — APPROVED** [review]
